sophia/src/parser/xml.rs

```diff
@@ -0,0 +1,1154 @@
+impl<F: TermFactory> Clone for Scope<F> {
+    fn clone(&self) -> Self {
+        Self {
+            ns: self.ns.clone(),
+            default: self.default.clone(),
+            base: self.base.clone(),
+            factory: self.factory.clone(),
+            datatype: self.datatype.clone(),
+            lang: self.lang.clone(),
+            text: self.text.clone(),
+            li: AtomicU64::new(self.li.load(Ordering::Relaxed)),
+        }
+    }
+}
+
+impl<F: TermFactory> Scope<F> {
+    /// Create a new `Scope` with the given term factory.
+    fn with_factory(f: F) -> Self {
+        Self::with_factory_rc(Rc::new(RefCell::new(f)))
+    }
+
+    /// Create a new `Scope` from a shared smartpointer to a term factory.
+    fn with_factory_rc(f: Rc<RefCell<F>>) -> Self {
+        let mut scope = Self {
+            ns: HashMap::new(),
+            default: None,
+            base: None,
+            factory: f,
+            datatype: None,
+            lang: None,
+            text: None,
+            li: AtomicU64::new(1),
+        };
+        // These namespaces are always in scope
+        scope
+            .add_prefix("xml", "http://www.w3.org/XML/1998/namespace#")
+            .unwrap();
+        scope
+            .add_prefix("xmlms", "https://www.w3.org/2000/xmlns/")
+            .unwrap();
+        scope
+    }
+
+    /// Add a new XML prefix to the namespace mapping.
+    fn add_prefix(&mut self, prefix: &str, value: &str) -> Result<()> {
+        if prefix == "_" {
+            panic!("reserved prefix")
+        } else {
+            let mut f = self.factory.borrow_mut();
+            self.ns.insert(
+                String::from(prefix),
+                Namespace::new(f.get_term_data(value))?,
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Set the default XML prefix.
+    fn set_default(&mut self, default: &str) -> Result<()> {
+        let mut f = self.factory.borrow_mut();
+        self.default = Some(Namespace::new(f.get_term_data(default))?);
+        Ok(())
+    }
+
+    /// Set the base IRI prefix.
+    fn set_base(&mut self, base: &str) -> Result<()> {
+        let mut f = self.factory.borrow_mut();
+        self.base = Some(Namespace::new(f.get_term_data(base))?);
+        Ok(())
+    }
+
+    fn set_datatype(&mut self, datatype: &str) -> Result<()> {
+        self.datatype = Some(self.expand_iri(datatype)?);
+        Ok(())
+    }
+
+    fn set_text<T: Into<Option<String>>>(&mut self, text: T) {
+        self.text = text.into();
+    }
+
+    fn expand_attribute(&self, attr: &str) -> Result<Term<F::TermData>> {
+        if let Some(separator_idx) = attr.chars().position(|c| c == ':') {
+            let prefix = &attr[..separator_idx];
+            let reference = &attr[separator_idx + 1..];
+            if let Some(ns) = self.ns.get(prefix) {
+                ns.get(self.factory.borrow_mut().get_term_data(reference))
+            } else {
+                panic!("unknown namespace: {}", prefix)
```

Comment:
return `Err()` instead of `panic!` ... (apply many time in this file)

Reply:
Yes I know, this is WIP, and ultimately all errors will be added to `XmlParser.triples`. I will need to edit the current errors to allow wrapping XML and decoding errors, but that may even be easier with `failure` (as I suggested in #8).